### PR TITLE
[MODGOBI-209] - Update location translation for central ordering

### PR DESCRIPTION
**Pull request by imerabishvili**
## Purpose
https://folio-org.atlassian.net/browse/MODGOBI-209

## Approach
Use different locations endpoint to obtain tenantId's as well.

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/rest/ResourcePaths.java [hunk 1]
```diff
@@ -12,7 +12,7 @@ private ResourcePaths() {
   public static final String ACQUISITION_METHOD_ENDPOINT = "/orders/acquisition-methods";
   public static final String GET_ORGANIZATION_ENDPOINT = "/organizations-storage/organizations";
   public static final String ACQUISITION_UNIT_ENDPOINT = "/acquisitions-units/units";
-  public static final String LOCATIONS_ENDPOINT = "/locations";
+  public static final String LOCATIONS_ENDPOINT = "/search/consortium/locations";
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/rest/ResourcePaths.java:15]: I suppose this endpoint will not work in non-ECS env, so we need to switch between new and one endpoints

File: src/main/java/org/folio/gobi/LookupService.java [hunk 2]
```diff
@@ -68,19 +69,26 @@ public LookupService(RestClient restClient) {
    * @param location
    * @return UUID of the location
    */
-  public CompletableFuture<String> lookupLocationId(String location) {
+  public CompletableFuture<LocationTranslationResult> lookupLocationId(String location) {
     logger.debug("lookupLocationId:: Trying to look up locationId by location '{}'", location);
-    String query = HelperUtils.encodeValue(String.format(CQL_CODE_STRING_FMT, location));
-    String endpoint = String.format(LOCATIONS_ENDPOINT + QUERY, query);
-    return restClient.handleGetRequest(endpoint).toCompletionStage().toCompletableFuture()
+    return restClient.handleGetRequest(LOCATIONS_ENDPOINT).toCompletionStage().toCompletableFuture()
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/gobi/LookupService.java:74]: I suppose limit param will be 10 if we not specify it, we need to check what max limit is supported by mod-search and use it

File: src/main/java/org/folio/rest/ResourcePaths.java (hunk 1 above)
Comment (SerhiiNosko) [reply src/main/java/org/folio/rest/ResourcePaths.java:15]: This is query that can help to determine do we need to use /search/consortium/locations endpoint
/orders-storage/settings?limit=1&query=key=ALLOW_ORDERING_WITH_AFFILIATED_LOCATIONS

Comment (imerabishvili) [reply src/main/java/org/folio/rest/ResourcePaths.java:15]: I tested and the approach works for snapshot environment as well. Can you please double check?

Comment (SerhiiNosko) [reply src/main/java/org/folio/rest/ResourcePaths.java:15]: yes, it is working on non-ecs as well, good to know, thanks
image

File: src/main/java/org/folio/gobi/LookupService.java (hunk 2 above)
Comment (SerhiiNosko) [reply src/main/java/org/folio/gobi/LookupService.java]: I see, 1000 is max
{
    "errors": [
        {
            "message": "getConsortiumLocations.limit must be less than or equal to 1000",
            "type": "ConstraintViolationException",
            "code": "validation_error",
            "parameters": []
        }
    ],
    "total_records": 1
}

File: src/main/java/org/folio/gobi/LookupService.java [hunk 3]
```diff
@@ -68,19 +69,26 @@ public LookupService(RestClient restClient) {
...
+    return restClient.handleGetRequest(LOCATIONS_ENDPOINT + "?limit=1000").toCompletionStage().toCompletableFuture()
       .thenCompose(locations -> {
-        String locationId = HelperUtils.extractLocationId(locations);
-        if (StringUtils.isEmpty(locationId)) {
-          logger.warn("lookupLocationId:: Location '{}' not found", location);
-          return completedFuture(null);
+        JsonArray jsonArray = locations.getJsonArray("locations");
+        for (int i = 0; i < jsonArray.size(); i++) {
+          JsonObject locationObject = jsonArray.getJsonObject(i);
+          if (location != null && location.equals(locationObject.getString("code"))) {
```
Comment (azizbekxm) [inline src/main/java/org/folio/gobi/LookupService.java:79]: let's think about two edge cases here.
1. What is location == null. Seems there is no need to work when location null, so we can do return  `return completedFuture(null);` in the beginning before send request to mod-search.
2. What is jsonArray is empty. then we can't get locationObject. we can also do condition for this

File: src/main/java/org/folio/gobi/LookupService.java (hunk 2 above)
Comment (SerhiiNosko) [reply src/main/java/org/folio/gobi/LookupService.java]: We need at least 10000, maybe if we not specify limit param than we will get all data, need to check it by inserting 10k records of locations

Comment (SerhiiNosko) [reply src/main/java/org/folio/gobi/LookupService.java]: looking into code of mod-search, if we not provide limit param - ES will set -1, so grab all data, but need to check via API call for example to fetch 2000k locations

File: src/main/java/org/folio/gobi/LookupService.java (hunk 3 above)
Comment (imerabishvili) [reply src/main/java/org/folio/gobi/LookupService.java:79]: 1. Location will not be null when called from FE. With this check just avoided NPE when working from swagger.
2. For empty jsonArray for cycle won't be executed so code will return correct result - null.

Comment (azizbekxm) [reply src/main/java/org/folio/gobi/LookupService.java:79]: That is make sense. I got it

File: src/main/java/org/folio/gobi/LookupService.java [hunk 4]
```diff
@@ -68,19 +69,26 @@ public LookupService(RestClient restClient) {
    * @param location
    * @return UUID of the location
    */
-  public CompletableFuture<String> lookupLocationId(String location) {
+  public CompletableFuture<LocationTranslationResult> lookupLocationId(String location) {
     logger.debug("lookupLocationId:: Trying to look up locationId by location '{}'", location);
-    String query = HelperUtils.encodeValue(String.format(CQL_CODE_STRING_FMT, location));
-    String endpoint = String.format(LOCATIONS_ENDPOINT + QUERY, query);
-    return restClient.handleGetRequest(endpoint).toCompletionStage().toCompletableFuture()
+    return restClient.handleGetRequest(LOCATIONS_ENDPOINT + "?limit=1000").toCompletionStage().toCompletableFuture()
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/gobi/LookupService.java:74]: limit can be deleted, default limit is 10000 will be used when this story will be completed: https://folio-org.atlassian.net/browse/MSEARCH-812

File: src/main/java/org/folio/rest/ResourcePaths.java (hunk 1 above)
Comment (SerhiiNosko) [reply src/main/java/org/folio/rest/ResourcePaths.java:15]: we will use default limit 10000 after this story completion https://folio-org.atlassian.net/browse/MSEARCH-812

File: src/main/java/org/folio/gobi/LookupService.java (hunk 4 above)
Comment (azizbekxm) [reply src/main/java/org/folio/gobi/LookupService.java]: removed

File: src/main/java/org/folio/gobi/LookupService.java [hunk 5]
```diff
@@ -62,28 +63,33 @@ public LookupService(RestClient restClient) {
...
+   * @return UUID of the location object
    */
-  public CompletableFuture<String> lookupLocationId(String location) {
-    logger.debug("lookupLocationId:: Trying to look up locationId by location '{}'", location);
-    String query = HelperUtils.encodeValue(String.format(CQL_CODE_STRING_FMT, location));
-    String endpoint = String.format(LOCATIONS_ENDPOINT + QUERY, query);
-    return restClient.handleGetRequest(endpoint).toCompletionStage().toCompletableFuture()
+  public CompletableFuture<LocationTranslationResult> lookupLocationId(String locationCode) {
+    logger.debug("lookupLocationId:: Trying to look up locationId by location '{}'", locationCode);
+    return restClient.handleGetRequest(LOCATIONS_ENDPOINT).toCompletionStage().toCompletableFuture()
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/gobi/LookupService.java:75]: please check if module permission is added to call mod-search enpoint

Comment (azizbekxm) [reply src/main/java/org/folio/gobi/LookupService.java:75]: Good catch, thanks. added permission and required module as well

Comment (SerhiiNosko) [reply src/main/java/org/folio/gobi/LookupService.java:75]: inventory-storage.locations.collection.get - and this one probably is not needed

Comment (azizbekxm) [reply src/main/java/org/folio/gobi/LookupService.java:75]: removed this one

**Review by SerhiiNosko — APPROVED** [review]

**Review by azizbekxm — APPROVED** [review]

**Comment by azizbekxm** [comment]
image

**Pull request merged**